flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ExecutionAttemptMappingProvider.java

```diff
@@ -38,32 +38,34 @@
     private final List<ExecutionVertex> tasks;
 
     /** The cached mapping, which would only be updated on miss. */
-    private final LinkedHashMap<ExecutionAttemptID, ExecutionVertex> cachedTasksById;
+    private volatile Map<ExecutionAttemptID, ExecutionVertex> cachedTasksById;
 
     public ExecutionAttemptMappingProvider(Iterable<ExecutionVertex> tasksIterable) {
         this.tasks = new ArrayList<>();
         tasksIterable.forEach(this.tasks::add);
 
-        this.cachedTasksById =
-                new LinkedHashMap<ExecutionAttemptID, ExecutionVertex>(tasks.size()) {
-
-                    @Override
-                    protected boolean removeEldestEntry(
-                            Map.Entry<ExecutionAttemptID, ExecutionVertex> eldest) {
-                        return size() > tasks.size();
-                    }
-                };
+        this.cachedTasksById = new ConcurrentHashMap<>(tasks.size());
     }
 
     public Optional<ExecutionVertex> getVertex(ExecutionAttemptID id) {
-        if (!cachedTasksById.containsKey(id)) {
-            cachedTasksById.putAll(getCurrentAttemptMappings());
-            if (!cachedTasksById.containsKey(id)) {
-                // the task probably gone after a restart
-                cachedTasksById.put(id, null);
+        ExecutionVertex vertex = cachedTasksById.get(id);
+        if (vertex != null) {
+            return Optional.of(vertex);
+        }
+
+        return updateAndGet(id);
+    }
+
+    private Optional<ExecutionVertex> updateAndGet(ExecutionAttemptID id) {
+        synchronized (tasks) {
```

Comment:
Why do you prefer synchronizing on variable over making the method `synchronized`? Is there a reason for that? Maybe I am missing something :thinking:? 

Personally I find it a bit misleading. (What was the reason for the choice of this variable, how is it connected with the synchronized block, where else it is used for synchronization etc.)

Reply:
Synchronization on `this` is always dangerous since it can lead to deadlock if somebody decides to synchronize on this object outside. So in my case, I can create the new private lock object or I can use some other private variable in this class. I decided to choose the second option because actually, it looks pretty reasonable for me - synchronization on tasks shows that only one thread can have access to tasks which is true.

Reply:
Ah, makes sense! Thanks for the explanation!